src/tools/rbd_mirror/ImageMap.cc

```diff
@@ -0,0 +1,613 @@
+                                                const Updates &remove) {
+  dout(20) << dendl;
+
+  for (auto const &update : remove) {
+    m_listener.remove_image(peer_uuid, update.global_image_id, update.instance_id);
+  }
+}
+
+template <typename I>
+void ImageMap<I>::handle_load(const std::map<std::string, cls::rbd::MirrorImageMap> &image_mapping) {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  m_policy->init(image_mapping);
+}
+
+template <typename I>
+void ImageMap<I>::schedule_add_action(const std::string &global_image_id) {
+  dout(20) << ": global_image_id=" << global_image_id << dendl;
+
+  Context *on_map = new SubActionMapContext(this, global_image_id);
+
+  Context *on_finish = new FunctionContext([this, global_image_id, on_map](int r) {
+      assert(m_lock.is_locked());
+      if (r < 0) {
+        derr << ": failed to map global_image_id=" << global_image_id << dendl;
+      }
+
+      delete on_map;
+    });
+
+  Policy::Action action = Policy::Action::create_add_action(on_map, on_finish);
+  append_and_schedule_action(global_image_id, action);
+}
+
+template <typename I>
+void ImageMap<I>::schedule_remove_action(const std::string &global_image_id) {
+  dout(20) << ": global_image_id=" << global_image_id << dendl;
+
+  Context *on_unmap = new SubActionUnmapContext(this, global_image_id);
+  Context *on_purge = new SubActionPurgeContext(this, global_image_id);
+
+  Context *on_finish = new FunctionContext(
+    [this, global_image_id, on_unmap, on_purge](int r) {
+      assert(m_lock.is_locked());
+      if (r < 0) {
+        derr << ": failed to unmap global_image_id=" << global_image_id << dendl;
+      }
+
+      if (!m_policy->actions_pending(global_image_id)) {
+        // release tracking states
+        m_peer_map.erase(global_image_id);
+        m_policy->release_state(global_image_id);
+      }
+
+      delete on_unmap;
+      delete on_purge;
+    });
+
+  Policy::Action action = Policy::Action::create_remove_action(
+    on_unmap, on_purge, on_finish);
+  append_and_schedule_action(global_image_id, action);
+}
+
+template <typename I>
+void ImageMap<I>::schedule_shuffle_action(const std::string &global_image_id, bool is_remove) {
+  assert(m_lock.is_locked());
+
+  dout(20) << ": global_image_id=" << global_image_id << dendl;
+
+  Context *on_unmap = new SubActionUnmapContext(this, global_image_id);
+  Context *on_map = new SubActionMapContext(this, global_image_id);
+
+  Context *on_finish = new FunctionContext(
+    [this, global_image_id, on_map, on_unmap](int r) {
+      assert(m_lock.is_locked());
+
+      if (r < 0) {
+        derr << ": failed to shuffle global_image_id=" << global_image_id << dendl;
+      }
+
+      delete on_map;
+      delete on_unmap;
+    });
+
+  Policy::Action action = Policy::Action::create_shuffle_action(
+    on_map, on_unmap, on_finish);
+
+  if (is_remove) {
+    prepend_and_schedule_action(global_image_id, action);
```

Comment:
The purpose of the prepend logic is to queue a shuffle action (due to instance removal) in the front the action queue. I think this is still required, right?

Reply:
If you had an action in-progress, you can't really interrupt it mid-state machine for a new action. If you are trying to send an acquire or release notification to an instance that was removed, the in-flight action can be aborted via the error path handling for the current state and it can move on to the next action. The policy shouldn't really allow the actions to get more than 2 deep at maximum since the last action can always be replaced by a new action if it hasn't started yet.